url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/model3d/TriangleNet3DComp.java

```diff
@@ -0,0 +1,2604 @@
+		HashSet<Segment3D> edg = new HashSet<Segment3D>();
+
+		Iterator<?> it = set.iterator();
+		TriangleElt3D trielt = null;
+		int cnt = 0;
+		double time = System.currentTimeMillis();
+		int size = set.size();
+		while (it.hasNext()) {
+			cnt++;
+			if (cnt % 10000 == 0) {
+
+				DB3DLogger.logger.log(Level.FINEST, "10.000 entries = "
+						+ (System.currentTimeMillis() - time) + "\n"
+						+ "Rest takes about: " + ((size - cnt) / 10000)
+						* (System.currentTimeMillis() - time));
+
+				time = System.currentTimeMillis();
+			}
+			trielt = (TriangleElt3D) it.next();
+			for (int j = 0; j < 3; j++) {
+				vert.add(trielt.getPoint(j));
+				edg.add(trielt.getSegment(j));
+			}
+		}
+		setVertices(vert.size());
+		setEdges(edg.size());
+	}
+
+	/*
+	 * Method to collect all elements starting at the given element.
+	 * 
+	 * @param set Set to collect elements
+	 * 
+	 * @param elt TriangleElt3D to start collection process
+	 */
+	private void makeSet(Set<Equivalentable> set, TriangleElt3D elt) {
+
+		LinkedList<TriangleElt3D> toVisit = new LinkedList<TriangleElt3D>();
+		TreeSet<Integer> checked = new TreeSet<Integer>();
+		toVisit.add(elt);
+		checked.add(elt.getID());
+
+		TriangleElt3D currTri;
+		TriangleElt3D nb;
+		// while still triangles "to visit":
+		while (!toVisit.isEmpty()) {
+			currTri = toVisit.pollFirst();
+			set.add(currTri);
+			for (int i = 0; i < 3; i++) {
+				nb = currTri.getNeighbour(i);
+				// if not already visited:
+				if (nb != null && !checked.contains(nb.getID())) {
+					toVisit.add(nb);
+					checked.add(nb.getID());
+				}
+			}
+		}
+
+	}
+
+	/**
+	 * Finds the outer boundary of a tin. Runs over all triangles and records
+	 * their outer segments.
+	 * 
+	 * @return HashSet with outer segments.
+	 * @throws IllegalStateException
+	 *             - if this TIN component is empty
+	 * @author Daria Golovko
+	 */
+	public Set<Segment3D> findTinBorder1() {
+		if (this.isEmpty())
+			throw new IllegalStateException("The TIN is empty!");
+
+		HashSet<Segment3D> outerSegments = new HashSet<Segment3D>();
+
+		TriangleElt3D tri = null;
+
+		Set s = this.getElementsViaSAM();
+
+		Iterator<Equivalentable> it = s.iterator();
+
+		while (it.hasNext()) {
+
+			// Searching through all triangles:
+
+			tri = (TriangleElt3D) it.next();
+
+			// If an external triangle was found:
+			for (int j = 0; j < 3; j++)
+				if (!tri.hasNeighbour(j))
+					outerSegments.add((Segment3D) tri.getSegment(j));
+		}
+
+		return outerSegments;
+	}
+
+	/**
+	 * Finds the outer boundary of a tin. Finds the fisrt outer triangle,
+	 * afterwards follows the edge of the tin.
+	 * 
+	 * @return HashSet with outer segments
+	 * @throws IllegalArgumentException
+	 *             - if index of a triangle point is not 0, 1 or 2. The
+	 *             exception originates in the method getPoint(int) of the class
+	 *             Triangle3D.
+	 * @author Daria Golovko
+	 */
+
+	public Set<Segment3D> findTinBorder2() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.